packages/graphing/controller/src/index.js

```diff
@@ -0,0 +1,338 @@
+import debug from 'debug';
+
+import cloneDeep from 'lodash/cloneDeep';
+import lodash from 'lodash';
+import isEqual from 'lodash/isEqual';
+import { tools } from '@pie-lib/graphing';
+
+const { utils } = tools;
+const { sinY, buildDataPoints, getAmplitudeAndFreq, FREQ_DIVIDER, parabolaFromTwoPoints } = utils;
+
+const log = debug('@pie-element:graphing:controller');
+
+export const equalPoint = (A, B) => {
+  // x1 = x2 & y1 = y2
+  let equalLabel = true;
+
+  if (A.label || B.label) {
+    equalLabel = isEqual(A.label, B.label);
+  }
+
+  return isEqual(A.x, B.x) && isEqual(A.y, B.y) && equalLabel;
+};
+
+export const equalSegment = (A, B, C, D) => {
+  // x1 = x3 & y1 = y3 & x2 = x4 & y2 = y4
+  return ((isEqual(A, C) && isEqual(B, D)) || ((isEqual(B, C) && isEqual(A, D))));
+};
+
+export const equalVector = (A, B, C, D) => {
+  // x1 = x3 & y1 = y3 & x2 = x4 & y2 = y4
+  return ((isEqual(A, C) && isEqual(B, D)));
+};
+
+export const equalLine = (A, B, C, D) => {
+  // (y2 - y1)/(x2 - x1) = (y4 - y3)/(x4 - x3);
+  return (((B.y - A.y) / (B.x - A.x)) === ((D.y - C.y) / (D.x - C.x)));
+};
+
+export const equalRay = (A, B, C, D) => {
+  // line & x1 = x3 & y1 = y3 & angle between (x1, y1) (x2, y2) is same as angle between (x3, y3) (x4, y4)
+  return ((
+        (B.y - A.y) / (B.x - A.x)
+      ) === (
+      (D.y - C.y) / (D.x - C.x))
+    ) && (A.x === C.x && A.y === C.y) &&
+    (
+      Math.atan2(B.y - A.y, B.x - A.x) * 180 / Math.PI ===
+      Math.atan2(D.y - C.y, D.x - C.x) * 180 / Math.PI
+    );
+};
+
+export const equalPolygon = (pointsA, pointsB) => {
+  const sessAnswerPoints = lodash.uniqWith(pointsA, isEqual);
+  const withoutDuplicates = lodash.uniqWith(pointsB, isEqual);
+  const sB = lodash.orderBy(sessAnswerPoints, ['x','y'], ['asc', 'asc']);
+  const sD = lodash.orderBy(withoutDuplicates, ['x','y'], ['asc', 'asc']);
+
+  return isEqual(sD, sB);
+};
+
+export const equalCircle = (A, B, C, D) => {
```

Comment:
why not use `{root, edge}` or `{center, edge}` so `isEqual(a,b){ ... } `?

Reply:
I'm not sure I understand. Anyway, to check the equality of a circle, it's not enough only to check if rootA equal with rootB and egdeA equals edgeB, because root: (0,0), edge: (1, 1) is same circle defined by root: (0, 0), edge: (-1, -1), the radius of the circle is important.